### PR TITLE
reactTodoForm

### PR DESCRIPTION
**Pull request by ymys1**
[DEMO LINK](https://ymys1.github.io/react_add-todo-form/)

### CODE REVIEW

File: src/App.tsx [hunk 1]
```diff
@@ -1,61 +1,119 @@
...
+          {(!addValidated && !currentTitle) && (
+            <span className="error">Please enter a title</span>
+          )}
         </div>
 
         <div className="field">
-          <select data-cy="userSelect">
-            <option value="0" disabled>Choose a user</option>
-          </select>
+          <select
```
Comment (Anna-Kuzmenko) [inline src/App.tsx:73]: you should add lable 'User: ' to this select

File: src/App.tsx [hunk 2]
```diff
@@ -1,61 +1,119 @@
...
+            <span className="error">Please choose a user</span>
+          )}
         </div>
 
-        <button type="submit" data-cy="submitButton">
+        <button
+          type="submit"
+          data-cy="submitButton"
+          onClick={() => {
+            handleAddButton();
```
Comment (Anna-Kuzmenko) [inline src/App.tsx:104]: this function should be in form (onSubmit)

**Review by Anna-Kuzmenko — CHANGES_REQUESTED** [review]
Good job! Move your onClick function to onSabmit in form and add labels

File: src/App.tsx [hunk 3]
```diff
@@ -1,61 +1,121 @@
+import { useState } from 'react';
 import './App.scss';
-
-// import usersFromServer from './api/users';
-// import todosFromServer from './api/todos';
+import usersFromServer from './api/users';
+import todosFromServer from './api/todos';
+import { TodoList } from './components/TodoList';
+import { Todo, User, TodosWithUser } from './types/Types';
```
Comment (SerhiiKirik) [inline src/App.tsx:6]: Add types to the PR
image

File: src/App.tsx [hunk 4]
```diff
@@ -1,61 +1,121 @@
...
+  const todos: Todo[] = todosFromServer;
+
+  function preparedTodosWithUsers(): TodosWithUser[] {
+    return (
+      todos.map(todo => ({
+        ...todo,
+        user: users.find(user => user.id === todo.userId) || null,
+      }))
+    );
+  }
```
Comment (SerhiiKirik) [inline src/App.tsx:19]: why do you need return here?
```suggestion
  function preparedTodosWithUsers(): TodosWithUser[] (
      todos.map(todo => ({
        ...todo,
        user: users.find(user => user.id === todo.userId) || null,
      )))
    );
  }
```

File: src/App.tsx [hunk 5]
```diff
@@ -1,61 +1,121 @@
...
+
+  const [selectedUser, setSelectedUser] = useState('');
+  const [currentTitle, setCurrentTitle] = useState('');
+  // eslint-disable-next-line max-len
+  const [todosWithUsers, setTodosWithUsers] = useState<TodosWithUser[]>(preparedTodosWithUsers());
```
Comment (SerhiiKirik) [inline src/App.tsx:24]: Just wrap instead use eslint disable
```suggestion
  const [
   todosWithUsers,
   setTodosWithUsers
 ] = useState<TodosWithUser[]>(preparedTodosWithUsers());
```

File: src/App.tsx [hunk 6]
```diff
@@ -1,61 +1,121 @@
...
+            userId: Number(selectedUser),
+            completed: false,
+            user: users.find(user => user.id === Number(selectedUser)) || null,
+          },
+        ],
+      );
+      setAddValidated(true);
+      setSelectedUser('');
+      setCurrentTitle('');
+    } else {
```
Comment (SerhiiKirik) [inline src/App.tsx:46]: try to avoit `if / esle ` construction

File: src/App.tsx [hunk 7]
```diff
@@ -1,61 +1,121 @@
...
+          {
+            id: (todos.sort((a, b) => b.id - a.id)[0].id + 1),
+            title: currentTitle,
+            userId: Number(selectedUser),
+            completed: false,
+            user: users.find(user => user.id === Number(selectedUser)) || null,
+          },
+        ],
+      );
+      setAddValidated(true);
```
Comment (SerhiiKirik) [inline src/App.tsx:43]: make a handle reset fucntion

File: src/App.tsx [hunk 8]
```diff
@@ -1,61 +1,121 @@
...
+
+  const validatedTitleAndUser = currentTitle && selectedUser;
+
+  const handleAddButton = (event:React.FormEvent<HTMLInputElement>) => {
+    if (validatedTitleAndUser) {
+      event.preventDefault()
+      setTodosWithUsers(
+        [...todosWithUsers,
+          {
+            id: (todos.sort((a, b) => b.id - a.id)[0].id + 1),
```
Comment (SerhiiKirik) [inline src/App.tsx:35]: split this with variables. Also, what is `a` & `b`?

File: src/App.tsx [hunk 9]
```diff
@@ -1,61 +1,121 @@
...
+            title: currentTitle,
+            userId: Number(selectedUser),
+            completed: false,
+            user: users.find(user => user.id === Number(selectedUser)) || null,
```
Comment (SerhiiKirik) [inline src/App.tsx:39]: create a separate function for it & reuse

File: src/App.tsx [hunk 10]
```diff
@@ -1,61 +1,121 @@
...
+
   return (
     <div className="App">
       <h1>Add todo form</h1>
 
-      <form action="/api/users" method="POST">
+      <form
+        action="/api/users"
+        method="POST"
+        onSubmit={(event) => (handleAddButton(event))}
```
Comment (SerhiiKirik) [inline src/App.tsx:58]: it's the same
in all cases
```suggestion
        onSubmit={handleAddButton}
```

File: src/App.tsx [hunk 11]
```diff
@@ -1,61 +1,121 @@
...
+              type="text"
+              data-cy="titleInput"
+              placeholder="Enter a title"
+              value={currentTitle}
+              onChange={event => setCurrentTitle(event.target.value)}
+            />
+          </label>
+          {(!addValidated && !currentTitle) && (
+            <span className="error">Please enter a title</span>
+          )}
```
Comment (SerhiiKirik) [inline src/App.tsx:72]:
```suggestion
          <label htmlFor="titleInput">
          
            <input
              type="text"
              data-cy="titleInput"
              placeholder="Enter a title"
              value={currentTitle}
              onChange={event => setCurrentTitle(event.target.value)}
            />
          </label>
          
          {(!addValidated && !currentTitle) && (
            <span className="error">Please enter a title</span>
          )}
```

File: src/App.tsx [hunk 12]
```diff
@@ -1,61 +1,121 @@
...
+        user: users.find(user => user.id === todo.userId) || null,
+      }))
+    );
+  }
+
+  const [selectedUser, setSelectedUser] = useState('');
+  const [currentTitle, setCurrentTitle] = useState('');
+  // eslint-disable-next-line max-len
+  const [todosWithUsers, setTodosWithUsers] = useState<TodosWithUser[]>(preparedTodosWithUsers());
+  const [addValidated, setAddValidated] = useState(true);
```
Comment (SerhiiKirik) [inline src/App.tsx:25]:
```suggestion
  const [isValid, setIsValid] = useState(true);
```

Comment (SerhiiKirik) [inline src/App.tsx:25]:
```suggestion
  const [isValid, setIsValid] = useState(true);
```

Comment (SerhiiKirik) [inline src/App.tsx:25]:
```suggestion
  const [isValid, setIsValid] = useState(true);
```

**Review by SerhiiKirik — CHANGES_REQUESTED** [review]
Well done 🔥 
Let's fix some bugs 🧑‍💻 
I can add todo just with spaces in the title
image

**Comment by ymys1** [comment]
Dont know why, but make new commit(
 I dont know if it'll be comfortable to review if I'll throw corrected file and code here, if not - I'll ask about it on QnA, just tell me)
 `import { useState } from 'react';
import './App.scss';
import usersFromServer from './api/users';
import todosFromServer from './api/todos';
import { TodoList } from './components/TodoList';
import { Todo, User, TodosWithUser } from './types/Types';

export const App = () => {
  const users: User[] = usersFromServer;
  const todos: Todo[] = todosFromServer;

  const preparedTodosWithUsers = () => (
    todos.map(todo => ({
      ...todo,
      user: users.find(user => user.id === todo.userId) || null,
    }))
  );

  const [selectedUser, setSelectedUser] = useState('');
  const [currentTitle, setCurrentTitle] = useState('');
  const [
    todosWithUsers,
    setTodosWithUsers,
  ] = useState<TodosWithUser[]>(preparedTodosWithUsers());
  const [isValid, setIsValid] = useState(true);

  const resetValidated = () => {
    setIsValid(true);
  };

  const resetSelectedUser = () => {
    setSelectedUser('');
  };

  const resetCrrentTitle = () => {
    setCurrentTitle('');
  };

  const validatedTitleAndUser = currentTitle && selectedUser;

  const handleAddButton = (event:React.FormEvent<HTMLInputElement>) => {
    if (validatedTitleAndUser) {
      event.preventDefault();
      setTodosWithUsers(
        [...todosWithUsers,
          {
            id: (todos.sort((a, b) => b.id - a.id)[0].id + 1),
            title: currentTitle,
            userId: Number(selectedUser),
            completed: false,
            user: users.find(user => user.id === Number(selectedUser)) || null,
          },
        ],
      );
      resetValidated();
      resetSelectedUser();
      resetCrrentTitle();
    }

    if (!validatedTitleAndUser) {
      setIsValid(false);
    }
  };

  return (
    <div className="App">
      <h1>Add todo form</h1>

      <form
        action="/api/users"
        method="POST"
        onSubmit={() => handleAddButton}
      >
        <div className="field">
          <label htmlFor="titleInput">

            <input
              type="text"
              data-cy="titleInput"
              placeholder="Enter a title"
              value={currentTitle}
              onChange={event => setCurrentTitle(event.target.value.trim())}
            />
          </label>

          {(!isValid && !currentTitle) && (
            <span className="error">Please enter a title</span>
          )}
        </div>

        <div className="field">
          <label htmlFor="userSelect">
            <select
              data-cy="userSelect"
              value={selectedUser}
              onChange={event => setSelectedUser(event.target.value)}
            >
              <option
                value=""
                selected
                disabled
              >
                Choose a user
              </option>

              {users.map(user => (
                <option
                  key={user.id}
                  value={user.id}
                >
                  {user.name}
                </option>
              ))}
            </select>
          </label>
          {(!isValid && !selectedUser) && (
            <span className="error">Please choose a user</span>
          )}
        </div>

        <button
          type="submit"
          data-cy="submitButton"
        >
          Add
        </button>
      </form>

      {todosWithUsers.map(todo => (
        <TodoList
          key={todo.id}
          todo={todo}
        />
      ))}
    </div>
  );
};
`
[App.tsx.txt](https://github.com/mate-academy/react_add-todo-form/files/11439838/App.tsx.txt)

**Comment by ymys1** [comment]
Txt file is at the end of message
Sorry for that

**Review by SerhiiKirik — CHANGES_REQUESTED** [review]
Go to the Q&A

**Pull request closed without merging**